src/vs/platform/terminal/common/terminalEnvironment.ts

```diff
@@ -4,21 +4,110 @@
  *--------------------------------------------------------------------------------------------*/
 
 import { OperatingSystem, OS } from '../../../base/common/platform.js';
-import type { IShellLaunchConfig } from './terminal.js';
+import type { IShellLaunchConfig, TerminalShellType, PosixShellType, WindowsShellType, GeneralShellType } from './terminal.js';
 
 /**
  * Aggressively escape non-windows paths to prepare for being sent to a shell. This will do some
  * escaping inaccurately to be careful about possible script injection via the file path. For
  * example, we're trying to prevent this sort of attack: `/foo/file$(echo evil)`.
  */
-export function escapeNonWindowsPath(path: string): string {
+export function escapeNonWindowsPath(path: string, shellType?: TerminalShellType): string {
 	let newPath = path;
 	if (newPath.includes('\\')) {
 		newPath = newPath.replace(/\\/g, '\\\\');
 	}
-	const bannedChars = /[\`\$\|\&\>\~\#\!\^\*\;\<\"\']/g;
+
+	// Shell-specific escaping
+	switch (shellType) {
+		case PosixShellType.Bash:
+		case PosixShellType.Sh:
+		case PosixShellType.Zsh:
+		case WindowsShellType.GitBash:
+			return escapeForPosixShell(newPath);
+		case PosixShellType.Fish:
+			return escapeForFishShell(newPath);
+		case GeneralShellType.PowerShell:
+			// PowerShell should be handled separately in preparePathForShell
+			// but if we get here, use PowerShell escaping
+			return escapeForPowerShell(newPath);
+		default:
+			// Default to POSIX shell escaping for unknown shells
+			return escapeForPosixShell(newPath);
+	}
```

Comment:
These escape functions are quite verbose and repeat a lot. Could the implementation instead live after this switch statement, where inside the switch statement it defines the banned characters and what to replace with what?